server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java

```diff
@@ -121,24 +123,44 @@ public long getBoundingTileCount(MultiGeoValues.GeoValue geoValue, int precision
 
         @Override
         public int setValues(long[] values, MultiGeoValues.GeoValue geoValue, int precision) {
-            MultiGeoValues.BoundingBox bounds = geoValue.boundingBox();
+            return setValuesForCell(new int[] { 0, 0, 0 }, values, 0, precision, geoValue);
+        }
 
-            final double tiles = 1 << precision;
-            int minXTile = GeoTileUtils.getXTile(bounds.minX(), (long) tiles);
-            int minYTile = GeoTileUtils.getYTile(bounds.maxY(), (long) tiles);
-            int maxXTile = GeoTileUtils.getXTile(bounds.maxX(), (long) tiles);
-            int maxYTile = GeoTileUtils.getYTile(bounds.minY(), (long) tiles);
-            int idx = 0;
-            for (int i = minXTile; i <= maxXTile; i++) {
-                for (int j = minYTile; j <= maxYTile; j++) {
-                    Rectangle rectangle = GeoTileUtils.toBoundingBox(i, j, precision);
-                    if (geoValue.intersects(rectangle)) {
-                        values[idx++] = GeoTileUtils.longEncodeTiles(precision, i, j);
-                    }
+        private int setValuesForCell(int[] tile, long[] values, int valuesIndex, int precision, MultiGeoValues.GeoValue geoValue) {
+            Rectangle rectangle = GeoTileUtils.toBoundingBox(tile[0], tile[1], tile[2]);
+            GeoRelation relation = geoValue.relate(rectangle);
+            if (tile[2] == precision) {
+                if (GeoRelation.QUERY_DISJOINT != relation) {
+                    values[valuesIndex++] = GeoTileUtils.longEncodeTiles(tile[2], tile[0], tile[1]);
                 }
+                return valuesIndex;
             }
 
-            return idx;
+            if (GeoRelation.QUERY_INSIDE == relation) {
+                return setValuesForFullyContainedTile(tile, values, valuesIndex, precision);
+            }
+            if (GeoRelation.QUERY_CROSSES == relation) {
+                int[][] subTiles = GeoTileUtils.getSubTiles(tile[0], tile[1], tile[2]);
+                for (int[] subTile : subTiles) {
```

Comment:
I think we do not need to call that method and create the integer arrays, something like:

```
for (int I = 0; i < 2; i++) {
   for (int j = 0; j < 2; j++) {
       int nextX = 2 * x + i;
       int nextY = 2 * y + j;
       Rectangle r = GeoTileUtils.toBoundingBox(nextX, nextY, zoom + 1);
       valuesIndex = setValuesForCell(subTile, values, valuesIndex, precision, geoValue);
  }
}
```

WDYT?

Reply:
yeah! I am still getting used to the "do not create objects" way of writing code. I'll update. thanks!